core/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java

```diff
@@ -650,4 +641,62 @@ public String executor() {
     int getNumNodesConnected() {
         return connectedNodes.size();
     }
+
+    private static class ConnectedNodes implements Supplier<DiscoveryNode> {
+
+        // this classes uses both a set and a list to support faster operations. Insertion and contains for this class are O(1) thanks
+        // to the use of the set and removal is O(n) due to the arraylist. In order to support a round-robin scheme through the connected
+        // nodes, this class uses a counter and retrieves by index from the list. The arraylist enables us to do this in O(1).
+        private final Set<DiscoveryNode> nodeSet = new HashSet<>();
```

Comment:
I don't like to have 2 data structures that we have to keep in sync. if you really wanna optimize for removal / insertion we can just use an array or an array list and sort it? I doubt we should to be honest. Maybe we just go with a list and linear scan or we use a set only and the same iterator approach and with the synchronization in place we can just set the iterator to null if we changed the set? that might be the easiest